### PR TITLE
chore: remove result set from router

### PR DESCRIPTION
**Pull request by cisse21**
# Description

Remove result set to simplify the router code

## Notion Ticket

 [ Notion Link ](https://www.notion.so/rudderstacks/8aac9087df644365acdf64e28e290153?v=6e06b0a5ade24f0aa5ffe05dc2972e84&p=e3e55801a29b40a6932e4d4572b899a6&pm=s)

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: router/router.go [hunk 1]
```diff
@@ -397,9 +335,13 @@ func (worker *workerT) batch(routerJobs []types.RouterJobT) []types.DestinationJ
 
 func (worker *workerT) workerProcess() {
 	timeout := time.After(jobsBatchTimeout)
+	latestResultSetTime := time.Now()
```
Comment (atzoum) [inline router/router.go:338]:
```suggestion
```

Add a new workerT field instead

```
type workerT struct {
	latestAssignedTime         time.Time
}
```

File: router/router.go [hunk 2]
```diff
@@ -397,9 +335,13 @@ func (worker *workerT) batch(routerJobs []types.RouterJobT) []types.DestinationJ
...
 	for {
 		select {
 		case message, hasMore := <-worker.channel:
+			if latestResultSetTime != message.workerAssignedTime {
+				latestResultSetTime = message.workerAssignedTime
+			}
```
Comment (atzoum) [inline router/router.go:344]:
```suggestion
			worker.latestAssignedTime = message.workerAssignedTime
```

File: router/router.go [hunk 3]
```diff
@@ -450,19 +392,18 @@ func (worker *workerT) workerProcess() {
...
+				JobID:               job.JobID,
+				SourceID:            parameters.SourceID,
+				DestinationID:       parameters.DestinationID,
+				AttemptNum:          job.LastJobStatus.AttemptNum,
+				ReceivedAt:          parameters.ReceivedAt,
+				CreatedAt:           job.CreatedAt.Format(misc.RFC3339Milli),
+				FirstAttemptedAt:    firstAttemptedAt,
+				TransformAt:         parameters.TransformAt,
+				JobT:                job,
+				LatestResultSetTime: latestResultSetTime,
```
Comment (atzoum) [inline router/router.go:405]:
```suggestion
```

File: router/router.go [hunk 4]
```diff
@@ -631,21 +572,14 @@ func (worker *workerT) processDestinationJobs() {
...
-					worker.localResultSet.resultSetBeginTime = time.Now()
-					worker.localResultSet.timeAllotted = resultSet.timeAllotted
-				}
-
 				// Assuming twice the overhead - defensive: 30% was just fine though
 				// In fact, the timeout should be more than the maximum latency allowed by these workers.
 				// Assuming 10s maximum latency
-				elapsed := time.Since(worker.localResultSet.resultSetBeginTime)
-				threshold := time.Duration(2.0 * math.Max(float64(worker.localResultSet.timeAllotted), float64(10*time.Second)))
+				elapsed := time.Since(destinationJob.JobMetadataArray[0].LatestResultSetTime)
```
Comment (atzoum) [inline router/router.go:578]:
```suggestion
				elapsed := time.Since(worker.latestAssignedTime)
```

File: router/router.go [hunk 5]
```diff
@@ -631,21 +572,14 @@ func (worker *workerT) processDestinationJobs() {
...
+				threshold := worker.rt.routerTimeout
 				if elapsed > threshold {
+					fmt.Println("TimeStamp", destinationJob.JobMetadataArray[0].LatestResultSetTime)
```
Comment (atzoum) [inline router/router.go:581]:
```suggestion
					fmt.Println("TimeStamp", worker.latestAssignedTime)
```

File: router/types/types.go [hunk 6]
```diff
@@ -33,20 +33,19 @@ type DestinationJobT struct {
...
+	DestinationID       string          `json:"destinationId"`
+	AttemptNum          int             `json:"attemptNum"`
+	ReceivedAt          string          `json:"receivedAt"`
+	CreatedAt           string          `json:"createdAt"`
+	FirstAttemptedAt    string          `json:"firstAttemptedAt"`
+	TransformAt         string          `json:"transformAt"`
+	WorkspaceID         string          `json:"workspaceId"`
+	Secret              json.RawMessage `json:"secret"`
+	JobT                *jobsdb.JobT    `json:"jobsT"`
+	LatestResultSetTime time.Time       `json:"latestResultSetTime"`
```
Comment (atzoum) [inline router/types/types.go:48]:
```suggestion
```

File: router/router.go (hunk 1 above)
Comment (cisse21) [reply router/router.go]: Done

File: router/router.go (hunk 3 above)
Comment (cisse21) [reply router/router.go]: Done

File: router/router.go (hunk 4 above)
Comment (cisse21) [reply router/router.go]: Done

File: router/types/types.go (hunk 6 above)
Comment (cisse21) [reply router/types/types.go]: Done

File: router/router.go (hunk 5 above)
Comment (cisse21) [reply router/router.go]: Removed these.. added them during debugging but forgot to clean up

File: router/router.go (hunk 2 above)
Comment (cisse21) [reply router/router.go]: Done

File: router/router.go [hunk 7]
```diff
@@ -631,20 +569,11 @@ func (worker *workerT) processDestinationJobs() {
...
-
 				// Assuming twice the overhead - defensive: 30% was just fine though
 				// In fact, the timeout should be more than the maximum latency allowed by these workers.
 				// Assuming 10s maximum latency
-				elapsed := time.Since(worker.localResultSet.resultSetBeginTime)
-				threshold := time.Duration(2.0 * math.Max(float64(worker.localResultSet.timeAllotted), float64(10*time.Second)))
+				elapsed := time.Since(worker.latestAssignedTime)
+				threshold := worker.rt.routerTimeout
 				if elapsed > threshold {
 					respStatusCode = types.RouterTimedOutStatusCode
```
Comment (fracasula) [inline router/router.go:578]: I can't find a test that is supposed to test this bit. Do we have one?

Comment (cisse21) [reply router/router.go:578]: There aren't any right now... Will try to add a test case for this

File: router/router.go [hunk 8]
```diff
@@ -631,20 +569,11 @@ func (worker *workerT) processDestinationJobs() {
 				//			router_delivery_exceeded_timeout -> goes to zero
 				ch := worker.trackStuckDelivery()
 
-				resultSetID := destinationJob.JobMetadataArray[0].ResultSetID
```
Comment (atzoum) [inline router/router.go:634]:
```suggestion
				if worker.latestAssignedTime != destinationJob.JobMetadataArray[0].WorkerAssignedTime {
					worker.latestAssignedTime = destinationJob.JobMetadataArray[0].WorkerAssignedTime
					worker.processingStartTime = time.Now()
				}
```

File: router/router.go [hunk 9]
```diff
@@ -631,20 +569,11 @@ func (worker *workerT) processDestinationJobs() {
...
-					worker.localResultSet.resultSetBeginTime = time.Now()
-					worker.localResultSet.timeAllotted = resultSet.timeAllotted
-				}
-
 				// Assuming twice the overhead - defensive: 30% was just fine though
 				// In fact, the timeout should be more than the maximum latency allowed by these workers.
 				// Assuming 10s maximum latency
-				elapsed := time.Since(worker.localResultSet.resultSetBeginTime)
-				threshold := time.Duration(2.0 * math.Max(float64(worker.localResultSet.timeAllotted), float64(10*time.Second)))
+				elapsed := time.Since(worker.latestAssignedTime)
```
Comment (atzoum) [inline router/router.go:575]:
```suggestion
				elapsed := time.Since(worker.processingStartTime)
```

File: router/router.go [hunk 10]
```diff
@@ -198,9 +192,8 @@ type workerT struct {
 	abortedUserMutex           sync.RWMutex
 	jobCountsByDestAndUser     map[string]*destJobCountsT
 	throttledAtTime            time.Time
+	latestAssignedTime         time.Time
```
Comment (atzoum) [inline router/router.go:195]:
```suggestion
	latestAssignedTime         time.Time
	processingStartTime        time.Time
```

File: router/router.go [hunk 11]
```diff
@@ -400,6 +339,7 @@ func (worker *workerT) workerProcess() {
 	for {
 		select {
 		case message, hasMore := <-worker.channel:
+			worker.latestAssignedTime = message.workerAssignedTime
```
Comment (atzoum) [inline router/router.go:342]:
```suggestion
```

File: router/router.go [hunk 12]
```diff
@@ -460,9 +400,7 @@ func (worker *workerT) workerProcess() {
 				FirstAttemptedAt: firstAttemptedAt,
 				TransformAt:      parameters.TransformAt,
 				JobT:             job,
-				PickedAtTime:     message.workerAssignedTime,
-				ResultSetID:      message.resultSetID,
-				WorkspaceId:      parameters.WorkspaceID,
+				WorkspaceID:      parameters.WorkspaceID,
```
Comment (atzoum) [inline router/router.go:403]:
```suggestion
				WorkspaceID:      parameters.WorkspaceID,
				WorkerAssignedTime: message.workerAssignedTime,
```

File: router/types/types.go [hunk 13]
```diff
@@ -42,11 +41,9 @@ type JobMetadataT struct {
 	CreatedAt        string          `json:"createdAt"`
 	FirstAttemptedAt string          `json:"firstAttemptedAt"`
 	TransformAt      string          `json:"transformAt"`
-	WorkspaceId      string          `json:"workspaceId"`
+	WorkspaceID      string          `json:"workspaceId"`
 	Secret           json.RawMessage `json:"secret"`
 	JobT             *jobsdb.JobT    `json:"jobsT"`
-	PickedAtTime     time.Time       `json:"pickedAtTime"`
```
Comment (atzoum) [inline router/types/types.go:48]:
```suggestion
	WorkerAssignedTime time.Time       `json:"-"`
```

**Review by atzoum — COMMENTED** [review]
If we want to adhere to the old behaviour, we need to start the timer when we encounter the 1st job of the router's batch during the worker's `processDestinationJobs`.

My suggestions below are towards that direction

**Review by atzoum — APPROVED** [review]

File: router/router.go (hunk 7 above)
Comment (cisse21) [reply router/router.go]: Added a test for this

**Review by fracasula — APPROVED** [review]

**Pull request merged**